Add cron jobs (#934)
woodpecker-ci/woodpecker: docs/docs/20-usage/45-cron.md

# Cron

> To enable cron jobs you need at least have push access to the repo.

Reviewer: ```suggestion
To configure cron jobs you need at least push access to the repo.
```

Reviewer: Currently the UI only allows repo admins to access the settings area if I remember correctly. I saw that recently that this access is not matching the permissions defined by the router.

:::warning

> by default cron wont trigger any step in pipelines, as the default event branch filters it.  

Reviewer: ```suggestion
By default pipelines triggered by cron jobs wont execute any steps in pipelines, as they are not part of the default event filter and you explicitly need to set a `event: cron` filter.
```

Read more at: [pipeline-syntax#event](/docs/usage/pipeline-syntax#event)
:::

> 1. So to start add the event filter to all steps where you like to run:

Reviewer: ```suggestion

## Add a new cron job

1. To create a new cron job adjust your pipeline config(s) and add the event filter to all steps you would like to run with the cron job:
```

    ```diff
     pipeline:
       sync_locales:
         image: weblate_sync
         settings:
           url: example.com
           token:
             from_secret: weblate_token
    +    when:

>     +      event: cron

Reviewer: ```suggestion
    +      event: cron
    +      cron: "name of the cron job" # if you only want to execute this step by a specific cron job
```

    ```

> 
> 2. Create a new cron job at repo settings

Reviewer: ```suggestion
1. Create a new cron job in the repository settings:
```

    ![cron settings](./cron-settings.png)

>     Schedule syntax can be found at https://pkg.go.dev/github.com/robfig/cron?utm_source=godoc#hdr-CRON_Expression_Format.  
>     Examples: `@every 5m`, `@daily`, `0 30 * * * *` ...

Reviewer: ```suggestion
    The supported schedule syntax can be found at <https://pkg.go.dev/github.com/robfig/cron?utm_source=godoc#hdr-CRON_Expression_Format>. 
    
    Examples: `@every 5m`, `@daily`, `0 30 * * * *` ...

    :::info
    Woodpeckers cron syntax starts with seconds instead of minutes (used by most linux cron-jobs).  Example: "At minute 30 every hour" would be `0 30 * * * *` instead of `30 * * * *` 
    :::
```

# Cron

> To configure cron jobs you need at least push access to the repo.

Reviewer: ```suggestion
To configure cron jobs you need at least push access to the repository.
```

:::warning

> By default pipelines triggered by cron jobs wont execute any steps in pipelines, as they are not part of the default event filter and you explicitly need to set a `event: cron` filter.

Reviewer: Should we stick to `crons` instead of `cron jobs`?

Author: well I dont have a strong opinion on this one

Reviewer: I rename all exitencens in code to be `cron` so they are much easier to recognize IMO. For comments, descriptios and docs I left `cron job` sometimes as it was nice to read. Hope you are fine with that change.

Read more at: [pipeline-syntax#event](/docs/usage/pipeline-syntax#event)
:::

## Add a new cron job

1. To create a new cron job adjust your pipeline config(s) and add the event filter to all steps you would like to run with the cron job:

    ```diff
     pipeline:
       sync_locales:
         image: weblate_sync
         settings:
           url: example.com
           token:
             from_secret: weblate_token
    +    when:
    +      event: cron
    +      cron: "name of the cron job" # if you only want to execute this step by a specific cron job
    ```
1. Create a new cron job in the repository settings:

    ![cron settings](./cron-settings.png)

>     The supported schedule syntax can be found at <https://pkg.go.dev/github.com/robfig/cron?utm_source=godoc#hdr-CRON_Expression_Format>. 

Reviewer: ```suggestion
    The supported schedule syntax can be found at <https://pkg.go.dev/github.com/robfig/cron?utm_source=godoc#hdr-CRON_Expression_Format>. If you need general understanding of the cron syntax <https://crontab.guru/> is a good place to start and experiment.
```

    
    Examples: `@every 5m`, `@daily`, `0 30 * * * *` ...

    :::info

>     Woodpeckers cron syntax starts with seconds instead of minutes (used by most linux cron-jobs).  Example: "At minute 30 every hour" would be `0 30 * * * *` instead of `30 * * * *` 

Reviewer: ```suggestion
    Woodpeckers cron syntax starts with seconds instead of minutes (used by most linux cron-jobs).  
    
    Example: "At minute 30 every hour" would be `0 30 * * * *` instead of `30 * * * *` 
```
